### PR TITLE
Warn on YUL switch statement with only default statement

### LINKED ISSUE #4285: Yul: disallow switch statements with only a default case
It made sense having them before `if` was introduced. Since `if` is supported this doesn't make too much sense anymore and for the sake of simplifying code, this could make sense.

(From https://github.com/ethereum/solidity/pull/3464/commits/027620766b6f531d89c12497207fa08b0a3f99c2#r185976506)

Reason against this: transformation steps could map single-case switch statements into if statements and vice versa. This may be beneficial for code generators, which opt to not implement if statements or which have an optimised implementation for if statements.

-> make this into a warning

### PR DESCRIPTION
**Pull request by hrkrshnn**
Fixes https://github.com/ethereum/solidity/issues/4285

TODO
1. [x] Figure out a way to make isoltest work.

### CODE REVIEW

**Comment by chriseth** [comment]
What is the exact problem here?

**Comment by hrkrshnn** [comment]
@chriseth One problem is that in `test/libyul/yulOptimizerTests/wordSizeTransform/switch_5.yul`

```solidity
{
    switch calldataload(0)
    default { sstore(8, 9) }
}
/// etc...
```

`isoltest` encounters a warning and fails for the case. The rest of the problems are similar, the testing framework is not expecting a warning so it fails. What do you think is a good way around this?

**Comment by chriseth** [comment]
Maybe the optimizer tests should ignore such warnings?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 1]
```diff
@@ -1234,7 +1234,7 @@ void ProtoConverter::visit(CaseStmt const& _x)
 
 void ProtoConverter::visit(SwitchStmt const& _x)
 {
-	if (_x.case_stmt_size() > 0 || _x.has_default_block())
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp]: The original motivation for this condition was the following logic:

> If a switch statement has at least one case OR a default block, process it.

We can break this down into the following permutations for a switch statement
  1. at least one case + default
  2. no case + default
  3. at least one case + no default
  4. no case + no default

I believe this PR only throws a warning for `2` (i.e., no switch case present but default present). For this specific permutation, the fuzzer (without the proposed modification) would generate a test case such as the following:

```
let x
switch x
  default {}
```

Which is okay for the `develop` branch for the same reason that retaining `pc()` (please see https://github.com/ethereum/solidity/pull/9040#issuecomment-634731923) is okay. However, as we did in the `pc()` instance, we could apply this patch to `breaking` branch when we error on switch cases with only default. At that time, this logic makes complete sense because we should not generate a switch case with only a default case. Or, in other words, we should only generate a default case in conjunction with at least one case statement.

tl;dr: Please revert this change as fuzzing switch statements with only a default block is still relevant for the develop branch. Please consider applying this patch in a futute PR that will error on switch statements with only a default case :-)

Comment (hrkrshnn) [reply test/tools/ossfuzz/protoToYul.cpp]: @bshastry We don't plan to introduce errors for switch with only a default. But if warnings are fine, then I guess, this change can be reverted.

File: Changelog.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ Compiler Features:
  * SMTChecker: Support array ``length``.
  * SMTChecker: Support array ``push`` and ``pop``.
  * Add support for natspec comments on state variables.
+ * Yul: Raise warning for switch statements that only has a default and no other cases.
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Yul: Raise warning for switch statements that only have a default and no other cases.
```

File: test/libsolidity/syntaxTests/controlFlow/storageReturn/assembly/switch_fine.sol [hunk 3]
```diff
@@ -1,20 +1,14 @@
 contract C {
     struct S { bool f; }
     S s;
-    function f(uint256 a) internal pure returns (S storage c) {
-        assembly {
-            switch a
-            default { c_slot := s_slot }
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/controlFlow/storageReturn/assembly/switch_fine.sol:7]: Can you keep this? It is useful to know that the check also works for the "only default" case.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: test/libsolidity/syntaxTests/controlFlow/storageReturn/assembly/switch_fine.sol (hunk 3 above)
Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/controlFlow/storageReturn/assembly/switch_fine.sol:7]: I have separated it to a new file `test/libsolidity/syntaxTests/controlFlow/storageReturn/assembly/switch_only_default_warn.sol` It bothered me that the name has `_fine` but still issues a warning.

**Comment by leonardoalt** [comment]
Needs rebase

**Review by leonardoalt — DISMISSED** [review]

**Review by chriseth — DISMISSED** [review]

**Review by chriseth — APPROVED** [review]
